api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -433,16 +449,51 @@ private User initializeUserIfNeeded() {
   }
 
   @Override
-  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest){
+  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest) {
     verifyInvitationKey(invitationVerificationRequest.getInvitationKey());
     return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
   }
 
-  private void verifyInvitationKey(String invitationKey){
-    if(invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
+  private void verifyInvitationKey(String invitationKey) {
+    if (invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
       throw new BadRequestException(
-        "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+          "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+    }
+  }
+
+  /*
+   * This un-authed API method is limited such that we only allow contact email updates before the user has signed in
+   * with the newly created gsuite account. Once the user has logged in, they can change their contact email through
+   * the normal profile update process.
+   */
+  @Override
+  public ResponseEntity<Void> updateContactEmail(UpdateContactEmailRequest updateContactEmailRequest) {
+    User user = userDao.findUserByEmail(updateContactEmailRequest.getUsername());
+    if (user.getFirstSignInTime() != null) {
+      return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+    }
+    try {
+      InternetAddress email = new InternetAddress(updateContactEmailRequest.getContactEmail());
+      email.validate();
+    } catch (AddressException e) {
+      return ResponseEntity.badRequest().build();
+    }
+    user.setContactEmail(updateContactEmailRequest.getContactEmail());
+    userDao.save(user);
+    return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+  }
+
+  @Override
+  public ResponseEntity<Void> resendWelcomeEmail(ResendWelcomeEmailRequest resendRequest) {
+    com.google.api.services.admin.directory.model.User googleUser =
```

Comment:
This should have the same sign-in check. Otherwise any anonymous client could reset everyone's passwords, just knowing their username.

Reply:
Added below

Reply:
This has the same issue still, since it only checks after the password reset.

Reply:
Believe I've resolved this

Reply:
Will update this check to match using the gsuite user instead.